component/client/settings/ClientSetting.tsx

```diff
@@ -0,0 +1,131 @@
+import * as React from "react";
+import {Grid, Row} from "react-bootstrap";
+
+import {storage} from "../backgroundContext";
+
+interface ISettingProperty {
+  original: string;
+  updated: string;
+}
+
+interface IClientSettingState {
+  username: ISettingProperty;
+  dirty: boolean;
+}
+
+class ClientSetting extends React.Component<any, IClientSettingState> {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      dirty: false,
+      username: {
+        original: "",
+        updated: ""
+      }
+    };
+
+    storage.get("username").then((username) => {
+      this.setState({
+        username: {
+          original: username,
+          updated: username,
+        }
+      });
+    });
+
+    this.handleSubmit = this.handleSubmit.bind(this);
+    this.handleUsernameChange = this.handleUsernameChange.bind(this);
+    this.reloadSettings = this.reloadSettings.bind(this);
+    this.reset = this.reset.bind(this);
+  }
+
+  public render(): JSX.Element {
+    return(
+      <Grid>
+        <form id="settings" onSubmit={this.handleSubmit}>
+          <Row className="input-group">
+            <label>
+              Username:
+              <input type="text" className="form-control" value={this.state.username.updated} onChange={this.handleUsernameChange}/>
+            </label>
+          </Row>
+          <Row className="input-group" style={{marginTop: "10px"}}>
+            <input type="submit" value="Submit" className="btn btn-info" disabled={!this.state.dirty}/>
+            <button className="btn btn-danger" style={{marginLeft: "50px"}} onClick={this.reset}>Reset</button>
+          </Row>
+        </form>
+      </Grid>
+    );
+  }
+
+  public handleSubmit(event) {
+    event.preventDefault();
+    let promise;
+    for (const key in this.state) {
```

Comment:
I think there is a react-bootstrap component field that will handle validation for you. It can also provide user feedback (i.e. change the field box red if it is not valid). I think it would be better to use that than to pollute the submission logic.